irrd/server/http/event_stream.py

```diff
@@ -0,0 +1,299 @@
+    encoding = "text"
+    stream_follower = None
+    websocket = None
+
+    async def on_connect(self, websocket: WebSocket) -> None:
+        await websocket.accept()
+        host = websocket.client.host if websocket.client else "[unknown client]"
+        if not is_client_permitted(host, "server.http.event_stream_access_list"):
+            await websocket.close(code=WS_1008_POLICY_VIOLATION)
+            return
+
+        self.websocket = websocket
+        await self.send_header()
+
+    async def send_header(self):
+        journaled_sources = [name for name, settings in get_setting("sources").items() if settings.get("keep_journal")]
+        dh = DatabaseHandler(readonly=True)
+        query = DatabaseStatusQuery().sources(journaled_sources)
+        sources_created = {row["source"]: row["created"].isoformat() for row in dh.execute_query(query)}
+        dh.close()
+
+        await self.message_callback(
+            {
+                "message_type": "stream_status",
+                "streamed_sources": journaled_sources,
+                "last_reload_times": sources_created,
+            }
+        )
+
+    async def message_callback(self, message: Any):
+        assert self.websocket
+        await self.websocket.send_text(ujson.encode(message))
+
+    async def on_receive(self, websocket: WebSocket, data: Any) -> None:
+        host = websocket.client.host if websocket.client else "[unknown client]"
+        logger.debug(f"event stream {host}: received {data}")
+        try:
+            request = EventStreamSubscriptionRequest.parse_raw(data)
+        except pydantic.ValidationError as exc:
+            await websocket.send_json(
+                {
+                    "message_type": "invalid_request",
+                    "errors": exc.errors(),
+                }
+            )
+            await websocket.close(code=WS_1003_UNSUPPORTED_DATA)
+            return
+
+        if self.stream_follower:
+            await websocket.send_json(
+                {
+                    "message_type": "invalid_request",
+                    "errors": [{"msg": "The stream is already running, request ignored."}],
+                }
+            )
+            return
+
+        self.stream_follower = await AsyncEventStreamFollower.create(
+            host, request.after_global_serial, self.message_callback
+        )
+
+    async def on_disconnect(self, websocket: WebSocket, close_code: int) -> None:
+        if self.stream_follower:
+            await self.stream_follower.close()
+            self.stream_follower = None
+
+
+class EventStreamSubscriptionRequest(pydantic.main.BaseModel):
+    message_type: Literal["subscribe"]
+    event_type: Literal["rpsl"]
+    after_global_serial: Optional[int]
+
+
+class AsyncEventStreamFollower:
+    @classmethod
+    async def create(
+        cls, host: str, after_global_serial: Optional[int], callback: Callable
+    ) -> Optional["AsyncEventStreamFollower"]:
```

Comment:
maybe:
```suggestion
    ) -> Optional[Self]:
```
now that we have typing_extensions anyways